akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -484,10 +496,7 @@ private[http] object HttpServerBluePrint {
             case IllegalUriException(errorInfo) =>
               finishWithIllegalRequestError(StatusCodes.BadRequest, errorInfo)
 
-            case _: ServerTerminationDeadlineReached =>
-              val response = settings.terminationDeadlineExceededResponse
-              log.warning(s"Closing connection with [{}] response, termination deadline exceeded. Server is shutting down...", response.status)
-              emitErrorResponse(response)
+            case ex: ServerTerminationDeadlineReached => failStage(ex)
```

Comment:
The previous behavior is part of the problem described in #3209: in the controller error handler we try to handle errors gracefully by trying to converts errors into a response + completion. In the renderer that extra (and not asked-for) response is then stuck while a response entity is streamed out. That means there's (too?) little control from stages on top of the stack like the user-handler or the GracefulTerminationStage to effectively kill bring down a connection.

Reply:
Right - not as nice, but indeed perhaps better than having 2 responses for the same request floating around.